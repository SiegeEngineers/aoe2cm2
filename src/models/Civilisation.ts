import GameVersion from "../constants/GameVersion";
import DraftOption from "./DraftOption";

enum Name {
    BRITONS = "Britons",
    BYZANTINES = "Byzantines",
    CELTS = "Celts",
    CHINESE = "Chinese",
    FRANKS = "Franks",
    GOTHS = "Goths",
    JAPANESE = "Japanese",
    MONGOLS = "Mongols",
    PERSIANS = "Persians",
    SARACENS = "Saracens",
    TEUTONS = "Teutons",
    TURKS = "Turks",
    VIKINGS = "Vikings",

    AZTECS = "Aztecs",
    HUNS = "Huns",
    KOREANS = "Koreans",
    MAYANS = "Mayans",
    SPANISH = "Spanish",

    ITALIANS = "Italians",
    INCAS = "Incas",
    INDIANS = "Indians",
    MAGYARS = "Magyars",
    SLAVS = "Slavs",

    BERBERS = "Berbers",
    ETHIOPIANS = "Ethiopians",
    MALIANS = "Malians",
    PORTUGUESE = "Portuguese",

    BURMESE = "Burmese",
    KHMER = "Khmer",
    MALAY = "Malay",
    VIETNAMESE = "Vietnamese",

    BULGARIANS = "Bulgarians",
    CUMANS = "Cumans",
    LITHUANIANS = "Lithuanians",
    TATARS = "Tatars",

    BURGUNDIANS = "Burgundians",
    SICILIANS = "Sicilians",

    BOHEMIANS = "Bohemians",
    POLES = "Poles",
}

class Civilisation extends DraftOption {

    public static readonly BRITONS: Civilisation = new Civilisation(Name.BRITONS, GameVersion.AOK);
    public static readonly BYZANTINES: Civilisation = new Civilisation(Name.BYZANTINES, GameVersion.AOK);
    public static readonly CELTS: Civilisation = new Civilisation(Name.CELTS, GameVersion.AOK);
    public static readonly CHINESE: Civilisation = new Civilisation(Name.CHINESE, GameVersion.AOK);
    public static readonly FRANKS: Civilisation = new Civilisation(Name.FRANKS, GameVersion.AOK);
    public static readonly GOTHS: Civilisation = new Civilisation(Name.GOTHS, GameVersion.AOK);
    public static readonly JAPANESE: Civilisation = new Civilisation(Name.JAPANESE, GameVersion.AOK);
    public static readonly MONGOLS: Civilisation = new Civilisation(Name.MONGOLS, GameVersion.AOK);
    public static readonly PERSIANS: Civilisation = new Civilisation(Name.PERSIANS, GameVersion.AOK);
    public static readonly SARACENS: Civilisation = new Civilisation(Name.SARACENS, GameVersion.AOK);
    public static readonly TEUTONS: Civilisation = new Civilisation(Name.TEUTONS, GameVersion.AOK);
    public static readonly TURKS: Civilisation = new Civilisation(Name.TURKS, GameVersion.AOK);
    public static readonly VIKINGS: Civilisation = new Civilisation(Name.VIKINGS, GameVersion.AOK);

    public static readonly AZTECS: Civilisation = new Civilisation(Name.AZTECS, GameVersion.CONQUERORS);
    public static readonly HUNS: Civilisation = new Civilisation(Name.HUNS, GameVersion.CONQUERORS);
    public static readonly KOREANS: Civilisation = new Civilisation(Name.KOREANS, GameVersion.CONQUERORS);
    public static readonly MAYANS: Civilisation = new Civilisation(Name.MAYANS, GameVersion.CONQUERORS);
    public static readonly SPANISH: Civilisation = new Civilisation(Name.SPANISH, GameVersion.CONQUERORS);

    public static readonly ITALIANS: Civilisation = new Civilisation(Name.ITALIANS, GameVersion.FORGOTTEN);
    public static readonly INCAS: Civilisation = new Civilisation(Name.INCAS, GameVersion.FORGOTTEN);
    public static readonly INDIANS: Civilisation = new Civilisation(Name.INDIANS, GameVersion.FORGOTTEN);
    public static readonly MAGYARS: Civilisation = new Civilisation(Name.MAGYARS, GameVersion.FORGOTTEN);
    public static readonly SLAVS: Civilisation = new Civilisation(Name.SLAVS, GameVersion.FORGOTTEN);

    public static readonly BERBERS: Civilisation = new Civilisation(Name.BERBERS, GameVersion.AFRICAN_KINGDOMS);
    public static readonly ETHIOPIANS: Civilisation = new Civilisation(Name.ETHIOPIANS, GameVersion.AFRICAN_KINGDOMS);
    public static readonly MALIANS: Civilisation = new Civilisation(Name.MALIANS, GameVersion.AFRICAN_KINGDOMS);
    public static readonly PORTUGUESE: Civilisation = new Civilisation(Name.PORTUGUESE, GameVersion.AFRICAN_KINGDOMS);

    public static readonly BURMESE: Civilisation = new Civilisation(Name.BURMESE, GameVersion.RISE_OF_RAJAS);
    public static readonly KHMER: Civilisation = new Civilisation(Name.KHMER, GameVersion.RISE_OF_RAJAS);
    public static readonly MALAY: Civilisation = new Civilisation(Name.MALAY, GameVersion.RISE_OF_RAJAS);
    public static readonly VIETNAMESE: Civilisation = new Civilisation(Name.VIETNAMESE, GameVersion.RISE_OF_RAJAS);

    public static readonly BULGARIANS: Civilisation = new Civilisation(Name.BULGARIANS, GameVersion.DEFINITIVE_EDITION);
    public static readonly CUMANS: Civilisation = new Civilisation(Name.CUMANS, GameVersion.DEFINITIVE_EDITION);
    public static readonly LITHUANIANS: Civilisation = new Civilisation(Name.LITHUANIANS, GameVersion.DEFINITIVE_EDITION);
    public static readonly TATARS: Civilisation = new Civilisation(Name.TATARS, GameVersion.DEFINITIVE_EDITION);

    public static readonly BURGUNDIANS: Civilisation = new Civilisation(Name.BURGUNDIANS, GameVersion.LORDS_OF_THE_WEST);
    public static readonly SICILIANS: Civilisation = new Civilisation(Name.SICILIANS, GameVersion.LORDS_OF_THE_WEST);

    public static readonly BOHEMIANS: Civilisation = new Civilisation(Name.BOHEMIANS, GameVersion.DAWN_OF_THE_DUKES);
    public static readonly POLES: Civilisation = new Civilisation(Name.POLES, GameVersion.DAWN_OF_THE_DUKES);


    // DO NOT CHANGE THE ORDER OF ELEMENTS IN THIS ARRAY!!!
    // ONLY APPEND NEW CIVILISATIONS AT THE END!!!
    public static readonly ALL = [
        Civilisation.AZTECS,
        Civilisation.BERBERS,
        Civilisation.BRITONS,
        Civilisation.BURMESE,
        Civilisation.BYZANTINES,
        Civilisation.CELTS,
        Civilisation.CHINESE,
        Civilisation.ETHIOPIANS,
        Civilisation.FRANKS,
        Civilisation.GOTHS,
        Civilisation.HUNS,
        Civilisation.INCAS,
        Civilisation.INDIANS,
        Civilisation.ITALIANS,
        Civilisation.JAPANESE,
        Civilisation.KHMER,
        Civilisation.KOREANS,
        Civilisation.MAGYARS,
        Civilisation.MALAY,
        Civilisation.MALIANS,
        Civilisation.MAYANS,
        Civilisation.MONGOLS,
        Civilisation.PERSIANS,
        Civilisation.PORTUGUESE,
        Civilisation.SARACENS,
        Civilisation.SLAVS,
        Civilisation.SPANISH,
        Civilisation.TEUTONS,
        Civilisation.TURKS,
        Civilisation.VIETNAMESE,
        Civilisation.VIKINGS,
        Civilisation.BULGARIANS,
        Civilisation.CUMANS,
        Civilisation.LITHUANIANS,
        Civilisation.TATARS,
        Civilisation.BURGUNDIANS,
        Civilisation.SICILIANS,
        Civilisation.BOHEMIANS,
        Civilisation.POLES,
    ];

    public readonly gameVersion: GameVersion;

    private constructor(name: Name, gameVersion: GameVersion) {
        super(name);
        this.gameVersion = gameVersion;
    }

}

export default Civilisation;
export {Name};
