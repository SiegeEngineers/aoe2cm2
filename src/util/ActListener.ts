import {DraftEvent} from "../types/DraftEvent";
import PlayerEvent from "../models/PlayerEvent";
import {DraftsStore} from "../models/DraftsStore";
import {ValidationId} from "../constants/ValidationId";
import {Util} from "./Util";
import Player from "../constants/Player";
import ActionType, {actionTypeFromAction} from "../constants/ActionType";
import DraftViews from "../models/DraftViews";
import fs from "fs";
import {logger} from "./Logger";
import AdminEvent from "../models/AdminEvent";
import path from "path";

export class ActListener {
    readonly dataDirectory: string;

    constructor(dataDirectory: string) {
        this.dataDirectory = dataDirectory;
    }

    actListener(draftsStore: DraftsStore, draftId: string, validateAndApply: (draftId: string, message: DraftEvent) => ValidationId[], socket: SocketIO.Socket, roomHost: string, roomGuest: string, roomSpec: string, skipSourceValidation = false) {
        return (message: PlayerEvent, fn: (retval: any) => void) => {
            logger.info("Got act message: %s", JSON.stringify(message), {draftId});

            const assignedRole = Util.getAssignedRole(socket, roomHost, roomGuest);

            if (!skipSourceValidation) {
                if (assignedRole === Player.NONE) {
                    logger.warn("Discarding spectator message", {draftId});
                    socket.emit('message', 'You shall not act.');
                    return;
                }

                if (message.executingPlayer !== assignedRole) {
                    logger.warn("Discarding fake message", {draftId});
                    socket.emit('message', 'You shall not impersonate your opponent.');
                    return;
                }

                if (!draftsStore.has(draftId)) {
                    logger.warn("Draft does not exist", {draftId});
                    socket.emit('message', 'This draft does not exist.');
                    return;
                }
            }

            const civilisationsList = draftsStore.getDraftOrThrow(draftId).preset.options.slice();
            message = Util.setRandomDraftOptionIfNeeded(message, draftId, draftsStore, civilisationsList);
            logger.info("Augmented message: %s", JSON.stringify(message), {draftId});

            const validationErrors: ValidationId[] = validateAndApply(draftId, message);
            if (validationErrors.length === 0) {

                const draftViews = draftsStore.getDraftViewsOrThrow(draftId);

                socket.nsp
                    .in(roomHost)
                    .emit("playerEvent", draftViews.getLastEventForHost());
                socket.nsp
                    .in(roomGuest)
                    .emit("playerEvent", draftViews.getLastEventForGuest());
                socket.nsp
                    .in(roomSpec)
                    .emit("playerEvent", draftViews.getLastEventForSpec());
                fn({status: 'ok', validationErrors});

                let adminEventCounter = 0;
                while (this.nextActionIsAdminEvent(draftsStore, draftId, adminEventCounter)) {
                    adminEventCounter++;
                    this.scheduleAdminEvent(adminEventCounter, draftsStore, draftId, draftViews, socket, roomHost, roomGuest, roomSpec);
                }
                if (draftViews.shouldRestartOrCancelCountdown()) {
                    draftsStore.restartOrCancelCountdown(draftId, this.dataDirectory);
                }
                this.finishDraftIfNoFurtherActions(draftViews, socket, draftsStore, draftId, roomHost, roomGuest, roomSpec);
            } else {
                logger.info("Validation yielded errors: %s", JSON.stringify(validationErrors), {draftId});
                fn({status: 'error', validationErrors});
            }
        };
    }

    nextActionIsAdminEvent(draftsStore: DraftsStore, draftId: string, offset: number) {
        const expectedActions = draftsStore.getExpectedActions(draftId, offset);
        if (expectedActions.length === 1) {
            return expectedActions[0].player === Player.NONE;
        }
        return false;
    }

    scheduleAdminEvent(adminEventCounter: number, draftsStore: DraftsStore, draftId: string, draftViews: DraftViews, socket: SocketIO.Socket, roomHost: string, roomGuest: string, roomSpec: string) {
        const expectedActions = draftsStore.getExpectedActions(draftId, adminEventCounter - 1);
        const expectedAction = expectedActions[0];
        if (expectedAction.player === Player.NONE) { // Admin Event
            if (actionTypeFromAction(expectedAction.action) === ActionType.REVEAL) {
                const draftEvent = new AdminEvent(expectedAction.player, expectedAction.action);
                setTimeout(() => {
                    logger.info('Executing admin event: %s', JSON.stringify(draftEvent), {draftId});
                    draftViews.reveal(expectedAction.action);
                    socket.nsp
                        .in(roomHost)
                        .emit("adminEvent", {
                            ...expectedAction,
                            events: draftViews.getHostDraft().events
                        });
                    socket.nsp
                        .in(roomGuest)
                        .emit("adminEvent", {
                            ...expectedAction,
                            events: draftViews.getGuestDraft().events
                        });
                    socket.nsp
                        .in(roomSpec)
                        .emit("adminEvent", {
                            ...expectedAction,
                            events: draftViews.getSpecDraft().events
                        });
                    draftsStore.addDraftEvent(draftId, draftEvent);
                    draftsStore.restartOrCancelCountdown(draftId, this.dataDirectory);
                    this.finishDraftIfNoFurtherActions(draftViews, socket, draftsStore, draftId, roomHost, roomGuest, roomSpec);
                }, adminEventCounter * 2000);
            } else {
                throw new Error("Unknown expected action! " + expectedAction);
            }
        }
    }

    finishDraftIfNoFurtherActions(draftViews: DraftViews, socket: SocketIO.Socket, draftsStore: DraftsStore,
                                  draftId: string, roomHost: string, roomGuest: string, roomSpec: string) {
        if (!draftViews.getActualDraft().hasNextAction()) {
            const draft = draftsStore.getDraftOrThrow(draftId);
            draft.startTimestamp = 0;
            draft.hostConnected = false;
            draft.guestConnected = false;
            logger.info("Saving draft: %s", JSON.stringify(draft), {draftId});
            const draftPath = path.join(this.dataDirectory, `${draftId}.json`);
            fs.writeFile(draftPath, JSON.stringify(draft), (err) => {
                logger.info("No further action expected. Disconnecting clients.", {draftId});
                for (let room of [roomHost, roomGuest, roomSpec]) {
                    socket.nsp.in(room).clients((error: any, clientIds: string[]) => {
                        if (error) throw error;
                        for (let clientId of clientIds) {
                            socket.nsp.connected[clientId].disconnect(true);
                        }
                    });
                }
                if (err) throw err;
                logger.info(`Draft saved to ${draftPath}`, {draftId});
                draftsStore.finishDraft(draftId);
            });
        }
    }
}
