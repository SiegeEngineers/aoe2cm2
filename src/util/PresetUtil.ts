import Preset from "../models/Preset";
import {Util} from "./Util";
import fs from "fs";
import {logger} from "./Logger";
import path from "path";

export const PresetUtil = {
    newPresetId(): string {
        return Util.newDraftId();
    },

    presetExists(presetId: string): boolean {
        const path = `presets/${presetId}.json`;
        return fs.existsSync(path);
    },

    createPreset(preset: Preset, presetDirectory: string) {
        let presetId = PresetUtil.newPresetId();
        while (PresetUtil.presetExists(presetId)) {
            presetId += Util.randomChar();
        }
        preset.presetId = presetId;
        const presetPath = path.join(presetDirectory,`${presetId}.json`);
        fs.writeFile(presetPath, JSON.stringify(preset), (err) => {
            if (err) throw err;
            logger.info(`Preset saved to ${presetPath}`, {presetId});
        });
        return presetId;
    },

    getIdFromUrl(): string | undefined {
        const match: RegExpMatchArray | null = window.location.pathname.match(/\/preset\/([A-Za-z0-9_]+)\/?.*/);
        if (match !== null) {
            return match[1];
        }
        alert('Could not get preset ID from url');
        return undefined;
    },

};
