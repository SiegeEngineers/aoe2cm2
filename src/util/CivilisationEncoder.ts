import Civilisation from "../models/Civilisation";

export const CivilisationEncoder = {
    encodeCivilisationArray(civilisations: Civilisation[]): string {
        let encoded: number = 0;
        for (let i = 0; i < Civilisation.ALL.length; i++) {
            if (civilisations.includes(Civilisation.ALL[i])) {
                encoded += Math.pow(2, i);
            }
        }
        return CivilisationEncoder.toHexString(encoded);
    },

    decodeCivilisationArray(encoded: string): Civilisation[] {
        if (encoded === '7fffffffff') {
            const civilisations = [...Civilisation.ALL];
            civilisations.sort((a, b) => a.name.localeCompare(b.name));
            return civilisations;
        }
        try {
            const encodedNumber = parseInt(encoded, 16);
            const binaryString: string = encodedNumber.toString(2);
            const bits: string[] = binaryString.split('');
            const civilisations = [];
            for (let i = 0; i < bits.length; i++) {
                if (bits[i] === '1') {
                    civilisations.push(Civilisation.ALL[bits.length - 1 - i]);
                }
            }
            civilisations.sort((a, b) => a.name.localeCompare(b.name));
            return civilisations;
        } catch (e) {
            console.log("Could not decode value as hex string: ", encoded);
            return [];
        }
    },

    toHexString(input: number): string {
        return input.toString(16);
    }
};
