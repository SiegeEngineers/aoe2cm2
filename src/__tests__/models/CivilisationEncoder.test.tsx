import {CivilisationEncoder} from "../../util/CivilisationEncoder";
import Civilisation from "../../models/Civilisation";

it('encode empty yields 0', () => {
    const encoded = CivilisationEncoder.encodeCivilisationArray([]);
    expect(encoded).toEqual(CivilisationEncoder.toHexString(0));
    expect(encoded).toEqual('0');
});

it('aztecs only yields 1', () => {
    const encoded = CivilisationEncoder.encodeCivilisationArray([Civilisation.AZTECS]);
    expect(encoded).toEqual(CivilisationEncoder.toHexString(1));
    expect(encoded).toEqual('1');
});

it('vikings only yields 2^30', () => {
    const encoded = CivilisationEncoder.encodeCivilisationArray([Civilisation.VIKINGS]);
    expect(encoded).toEqual(CivilisationEncoder.toHexString(Math.pow(2, 30)));
    expect(encoded).toEqual('40000000');
});

it('all 39 civs yield 2^39-1', () => {
    const encoded = CivilisationEncoder.encodeCivilisationArray(Civilisation.ALL);
    expect(Civilisation.ALL.length).toEqual(39);
    expect(encoded).toEqual(CivilisationEncoder.toHexString(Math.pow(2, 39) - 1));
    expect(encoded).toEqual('7fffffffff');
});

it('decode 0 yields empty array', () => {
    const decoded = CivilisationEncoder.decodeCivilisationArray('0');
    expect(decoded).toEqual([]);
});

it('decode 1 yields aztecs', () => {
    const decoded = CivilisationEncoder.decodeCivilisationArray('1');
    expect(decoded).toEqual([Civilisation.AZTECS]);

});

it('decode 2^30 yields vikings', () => {
    const decoded = CivilisationEncoder.decodeCivilisationArray('40000000');
    expect(decoded).toEqual([Civilisation.VIKINGS]);

});

it('decode 2^39-1 yields all civs', () => {
    const decoded = CivilisationEncoder.decodeCivilisationArray('7fffffffff');
    expect(decoded).toEqual([...Civilisation.ALL].sort((a, b) => a.name.localeCompare(b.name)));
});

it('decode invalid yields empty array', () => {
    const decoded = CivilisationEncoder.decodeCivilisationArray('invalid');
    expect(decoded).toEqual([]);
});

it('assert order of civilisations has not changed', () => {
    expect(Civilisation.ALL).toMatchSnapshot();
});

describe('The decoding shortcut does not lie', () => {
    it.each`
    civilisationArray | expectedLength
    ${'7ffffffff'}    | ${35}
    ${'1fffffffff'}   | ${37}
    ${'7fffffffff'}   | ${39}
  `('$civilisationArray', ({civilisationArray, expectedLength}) => {
        const decoded = CivilisationEncoder.decodeCivilisationArray(civilisationArray);
        expect(decoded.length).toEqual(expectedLength);
    });
});
