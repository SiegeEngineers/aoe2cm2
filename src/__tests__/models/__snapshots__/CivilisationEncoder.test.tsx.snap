// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assert order of civilisations has not changed 1`] = `
Array [
  Civilisation {
    "gameVersion": 2,
    "id": "Aztecs",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/aztecs-left.apng",
      "animated_right": "/images/units-animated/aztecs-right.apng",
      "emblem": "/images/civemblems/aztecs.png",
      "unit": "/images/civs/aztecs.png",
    },
    "name": "Aztecs",
  },
  Civilisation {
    "gameVersion": 4,
    "id": "Berbers",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/berbers-left.apng",
      "animated_right": "/images/units-animated/berbers-right.apng",
      "emblem": "/images/civemblems/berbers.png",
      "unit": "/images/civs/berbers.png",
    },
    "name": "Berbers",
  },
  Civilisation {
    "gameVersion": 1,
    "id": "Britons",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/britons-left.apng",
      "animated_right": "/images/units-animated/britons-right.apng",
      "emblem": "/images/civemblems/britons.png",
      "unit": "/images/civs/britons.png",
    },
    "name": "Britons",
  },
  Civilisation {
    "gameVersion": 5,
    "id": "Burmese",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/burmese-left.apng",
      "animated_right": "/images/units-animated/burmese-right.apng",
      "emblem": "/images/civemblems/burmese.png",
      "unit": "/images/civs/burmese.png",
    },
    "name": "Burmese",
  },
  Civilisation {
    "gameVersion": 1,
    "id": "Byzantines",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/byzantines-left.apng",
      "animated_right": "/images/units-animated/byzantines-right.apng",
      "emblem": "/images/civemblems/byzantines.png",
      "unit": "/images/civs/byzantines.png",
    },
    "name": "Byzantines",
  },
  Civilisation {
    "gameVersion": 1,
    "id": "Celts",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/celts-left.apng",
      "animated_right": "/images/units-animated/celts-right.apng",
      "emblem": "/images/civemblems/celts.png",
      "unit": "/images/civs/celts.png",
    },
    "name": "Celts",
  },
  Civilisation {
    "gameVersion": 1,
    "id": "Chinese",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/chinese-left.apng",
      "animated_right": "/images/units-animated/chinese-right.apng",
      "emblem": "/images/civemblems/chinese.png",
      "unit": "/images/civs/chinese.png",
    },
    "name": "Chinese",
  },
  Civilisation {
    "gameVersion": 4,
    "id": "Ethiopians",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/ethiopians-left.apng",
      "animated_right": "/images/units-animated/ethiopians-right.apng",
      "emblem": "/images/civemblems/ethiopians.png",
      "unit": "/images/civs/ethiopians.png",
    },
    "name": "Ethiopians",
  },
  Civilisation {
    "gameVersion": 1,
    "id": "Franks",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/franks-left.apng",
      "animated_right": "/images/units-animated/franks-right.apng",
      "emblem": "/images/civemblems/franks.png",
      "unit": "/images/civs/franks.png",
    },
    "name": "Franks",
  },
  Civilisation {
    "gameVersion": 1,
    "id": "Goths",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/goths-left.apng",
      "animated_right": "/images/units-animated/goths-right.apng",
      "emblem": "/images/civemblems/goths.png",
      "unit": "/images/civs/goths.png",
    },
    "name": "Goths",
  },
  Civilisation {
    "gameVersion": 2,
    "id": "Huns",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/huns-left.apng",
      "animated_right": "/images/units-animated/huns-right.apng",
      "emblem": "/images/civemblems/huns.png",
      "unit": "/images/civs/huns.png",
    },
    "name": "Huns",
  },
  Civilisation {
    "gameVersion": 3,
    "id": "Incas",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/incas-left.apng",
      "animated_right": "/images/units-animated/incas-right.apng",
      "emblem": "/images/civemblems/incas.png",
      "unit": "/images/civs/incas.png",
    },
    "name": "Incas",
  },
  Civilisation {
    "gameVersion": 3,
    "id": "Indians",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/indians-left.apng",
      "animated_right": "/images/units-animated/indians-right.apng",
      "emblem": "/images/civemblems/indians.png",
      "unit": "/images/civs/indians.png",
    },
    "name": "Indians",
  },
  Civilisation {
    "gameVersion": 3,
    "id": "Italians",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/italians-left.apng",
      "animated_right": "/images/units-animated/italians-right.apng",
      "emblem": "/images/civemblems/italians.png",
      "unit": "/images/civs/italians.png",
    },
    "name": "Italians",
  },
  Civilisation {
    "gameVersion": 1,
    "id": "Japanese",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/japanese-left.apng",
      "animated_right": "/images/units-animated/japanese-right.apng",
      "emblem": "/images/civemblems/japanese.png",
      "unit": "/images/civs/japanese.png",
    },
    "name": "Japanese",
  },
  Civilisation {
    "gameVersion": 5,
    "id": "Khmer",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/khmer-left.apng",
      "animated_right": "/images/units-animated/khmer-right.apng",
      "emblem": "/images/civemblems/khmer.png",
      "unit": "/images/civs/khmer.png",
    },
    "name": "Khmer",
  },
  Civilisation {
    "gameVersion": 2,
    "id": "Koreans",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/koreans-left.apng",
      "animated_right": "/images/units-animated/koreans-right.apng",
      "emblem": "/images/civemblems/koreans.png",
      "unit": "/images/civs/koreans.png",
    },
    "name": "Koreans",
  },
  Civilisation {
    "gameVersion": 3,
    "id": "Magyars",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/magyars-left.apng",
      "animated_right": "/images/units-animated/magyars-right.apng",
      "emblem": "/images/civemblems/magyars.png",
      "unit": "/images/civs/magyars.png",
    },
    "name": "Magyars",
  },
  Civilisation {
    "gameVersion": 5,
    "id": "Malay",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/malay-left.apng",
      "animated_right": "/images/units-animated/malay-right.apng",
      "emblem": "/images/civemblems/malay.png",
      "unit": "/images/civs/malay.png",
    },
    "name": "Malay",
  },
  Civilisation {
    "gameVersion": 4,
    "id": "Malians",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/malians-left.apng",
      "animated_right": "/images/units-animated/malians-right.apng",
      "emblem": "/images/civemblems/malians.png",
      "unit": "/images/civs/malians.png",
    },
    "name": "Malians",
  },
  Civilisation {
    "gameVersion": 2,
    "id": "Mayans",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/mayans-left.apng",
      "animated_right": "/images/units-animated/mayans-right.apng",
      "emblem": "/images/civemblems/mayans.png",
      "unit": "/images/civs/mayans.png",
    },
    "name": "Mayans",
  },
  Civilisation {
    "gameVersion": 1,
    "id": "Mongols",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/mongols-left.apng",
      "animated_right": "/images/units-animated/mongols-right.apng",
      "emblem": "/images/civemblems/mongols.png",
      "unit": "/images/civs/mongols.png",
    },
    "name": "Mongols",
  },
  Civilisation {
    "gameVersion": 1,
    "id": "Persians",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/persians-left.apng",
      "animated_right": "/images/units-animated/persians-right.apng",
      "emblem": "/images/civemblems/persians.png",
      "unit": "/images/civs/persians.png",
    },
    "name": "Persians",
  },
  Civilisation {
    "gameVersion": 4,
    "id": "Portuguese",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/portuguese-left.apng",
      "animated_right": "/images/units-animated/portuguese-right.apng",
      "emblem": "/images/civemblems/portuguese.png",
      "unit": "/images/civs/portuguese.png",
    },
    "name": "Portuguese",
  },
  Civilisation {
    "gameVersion": 1,
    "id": "Saracens",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/saracens-left.apng",
      "animated_right": "/images/units-animated/saracens-right.apng",
      "emblem": "/images/civemblems/saracens.png",
      "unit": "/images/civs/saracens.png",
    },
    "name": "Saracens",
  },
  Civilisation {
    "gameVersion": 3,
    "id": "Slavs",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/slavs-left.apng",
      "animated_right": "/images/units-animated/slavs-right.apng",
      "emblem": "/images/civemblems/slavs.png",
      "unit": "/images/civs/slavs.png",
    },
    "name": "Slavs",
  },
  Civilisation {
    "gameVersion": 2,
    "id": "Spanish",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/spanish-left.apng",
      "animated_right": "/images/units-animated/spanish-right.apng",
      "emblem": "/images/civemblems/spanish.png",
      "unit": "/images/civs/spanish.png",
    },
    "name": "Spanish",
  },
  Civilisation {
    "gameVersion": 1,
    "id": "Teutons",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/teutons-left.apng",
      "animated_right": "/images/units-animated/teutons-right.apng",
      "emblem": "/images/civemblems/teutons.png",
      "unit": "/images/civs/teutons.png",
    },
    "name": "Teutons",
  },
  Civilisation {
    "gameVersion": 1,
    "id": "Turks",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/turks-left.apng",
      "animated_right": "/images/units-animated/turks-right.apng",
      "emblem": "/images/civemblems/turks.png",
      "unit": "/images/civs/turks.png",
    },
    "name": "Turks",
  },
  Civilisation {
    "gameVersion": 5,
    "id": "Vietnamese",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/vietnamese-left.apng",
      "animated_right": "/images/units-animated/vietnamese-right.apng",
      "emblem": "/images/civemblems/vietnamese.png",
      "unit": "/images/civs/vietnamese.png",
    },
    "name": "Vietnamese",
  },
  Civilisation {
    "gameVersion": 1,
    "id": "Vikings",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/vikings-left.apng",
      "animated_right": "/images/units-animated/vikings-right.apng",
      "emblem": "/images/civemblems/vikings.png",
      "unit": "/images/civs/vikings.png",
    },
    "name": "Vikings",
  },
  Civilisation {
    "gameVersion": 6,
    "id": "Bulgarians",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/bulgarians-left.apng",
      "animated_right": "/images/units-animated/bulgarians-right.apng",
      "emblem": "/images/civemblems/bulgarians.png",
      "unit": "/images/civs/bulgarians.png",
    },
    "name": "Bulgarians",
  },
  Civilisation {
    "gameVersion": 6,
    "id": "Cumans",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/cumans-left.apng",
      "animated_right": "/images/units-animated/cumans-right.apng",
      "emblem": "/images/civemblems/cumans.png",
      "unit": "/images/civs/cumans.png",
    },
    "name": "Cumans",
  },
  Civilisation {
    "gameVersion": 6,
    "id": "Lithuanians",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/lithuanians-left.apng",
      "animated_right": "/images/units-animated/lithuanians-right.apng",
      "emblem": "/images/civemblems/lithuanians.png",
      "unit": "/images/civs/lithuanians.png",
    },
    "name": "Lithuanians",
  },
  Civilisation {
    "gameVersion": 6,
    "id": "Tatars",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/tatars-left.apng",
      "animated_right": "/images/units-animated/tatars-right.apng",
      "emblem": "/images/civemblems/tatars.png",
      "unit": "/images/civs/tatars.png",
    },
    "name": "Tatars",
  },
  Civilisation {
    "gameVersion": 7,
    "id": "Burgundians",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/burgundians-left.apng",
      "animated_right": "/images/units-animated/burgundians-right.apng",
      "emblem": "/images/civemblems/burgundians.png",
      "unit": "/images/civs/burgundians.png",
    },
    "name": "Burgundians",
  },
  Civilisation {
    "gameVersion": 7,
    "id": "Sicilians",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/sicilians-left.apng",
      "animated_right": "/images/units-animated/sicilians-right.apng",
      "emblem": "/images/civemblems/sicilians.png",
      "unit": "/images/civs/sicilians.png",
    },
    "name": "Sicilians",
  },
  Civilisation {
    "gameVersion": 8,
    "id": "Bohemians",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/bohemians-left.apng",
      "animated_right": "/images/units-animated/bohemians-right.apng",
      "emblem": "/images/civemblems/bohemians.png",
      "unit": "/images/civs/bohemians.png",
    },
    "name": "Bohemians",
  },
  Civilisation {
    "gameVersion": 8,
    "id": "Poles",
    "imageUrls": Object {
      "animated_left": "/images/units-animated/poles-left.apng",
      "animated_right": "/images/units-animated/poles-right.apng",
      "emblem": "/images/civemblems/poles.png",
      "unit": "/images/civs/poles.png",
    },
    "name": "Poles",
  },
]
`;
